### PR TITLE
Cahn-Hilliard-Navier-Stokes

### PR DESCRIPTION
**Pull request by PierreLaurentinCS**
# Description of the problem

- The last implementation of the Cahn-Hilliard-Navier-Stokes coupled equations were incorrectly reproducing classical test cases such as the rising bubble or the Rayleigh-Taylor instability
- The problem resided in the formulation of the Cahn-Hilliard equation and also major mistakes in the computation of physical properties in the coupling between Navier-Stokes and Cahn-Hilliard
- The thought process which eventually led to a change in the initial set of equation led to other additions. Among them : chemical potential smoothing, phase filtering (clipping and tanh) and other minor modifications 

# Description of the solution

- The Cahn-Hilliard equations were fixed and the use of mobility as physical property made it more readable and compact
- The coefficients multiplying the energies (bulk and interface) now depend on two parameters : epsilon (proportional to the size of the cell) and lambda which itself depends on the size of the cell and surface tension

# How Has This Been Tested?

- All the former tests were fixed to account for the new assembly : mms2d and 3d, cahn-hilliard-navier-stokes coupling, time-dependant boundary condition
- A new test was added to test the phase filtering
- Many canonical tests were performed. This should be illustrated soon with examples.

# Documentation

- Nothing was changed in the documentation, this will be the work of another PR

# Future changes

- During the development a gls stabilization was successfully implemented but it was not kept in this version because of the new assembly. This will be added in a future PR.
- The fixed angle of contact BC is not functional for extreme angles. New formulations will be tested in future PRs

### CODE REVIEW

File: source/solvers/cahn_hilliard.cc [hunk 1]
```diff
@@ -1185,6 +1230,81 @@ CahnHilliard<dim>::calculate_barycenter(const GlobalVectorType &solution,
...
+
+  const FEValuesExtractors::Scalar phase_order(0);
+
+  TrilinosWrappers::MPI::Vector filtered_solution_owned(
+    this->locally_owned_dofs, mpi_communicator);
+  filtered_solution_owned = this->present_solution;
+
+  filtered_solution.reinit(this->present_solution);
+
+  std::unordered_map<unsigned int, bool> filtered_cell_list;
```
Comment (blaisb) [inline source/solvers/cahn_hilliard.cc:1253]: Instead of yusing and std::unordered_map of bool use an unordered_set (which is exactly that in a way...)
If it's in the set it's true, if it's not it is false.

File: source/solvers/cahn_hilliard_assemblers.cc [hunk 2]
```diff
@@ -5,135 +5,86 @@
...
   StabilizedMethodsCopyData    &copy_data)
 {
   // Gather physical properties
-  const double well_height       = this->cahn_hilliard_parameters.well_height;
-  const double mobility_constant = this->mobility_constant;
-  const double epsilon           = scratch_data.epsilon;
+  const double epsilon   = this->epsilon;
+  const double cell_size = scratch_data.cell_size;
+  const double xi =
+    this->cahn_hilliard_parameters.potential_smoothing_coefficient;
```
Comment (blaisb) [inline source/solvers/cahn_hilliard_assemblers.cc:19]: Why is epsilon passed to the constructor explicitely but the smoothign parameter is still within the parameters? This should be explained a bit more above.

File: source/solvers/cahn_hilliard_assemblers.cc [hunk 3]
```diff
@@ -265,9 +172,11 @@ CahnHilliardAssemblerAngleOfContact<dim>::assemble_matrix(
...
 {
   if (!scratch_data.is_boundary_cell)
-    return;
-
-  const double epsilon = scratch_data.epsilon;
+    {
+      return;
+    }
+  const double epsilon = this->epsilon;
+  // std::cout<<"in angle of contact assembly"<<std::endl;
```
Comment (blaisb) [inline source/solvers/cahn_hilliard_assemblers.cc:179]:
```suggestion
```

**Review by blaisb — CHANGES_REQUESTED** [review]
FIrst batch of comments. Could you please add within the example folder the CHN examples you have, even if they are not documented, this way we can start keeping track of them. At least static bubble and rising bubble.

File: source/solvers/cahn_hilliard.cc [hunk 4]
```diff
@@ -44,30 +46,35 @@ CahnHilliard<dim>::setup_assemblers()
   this->assemblers.push_back(
     std::make_shared<CahnHilliardAssemblerAngleOfContact<dim>>(
       this->simulation_control,
+      this->simulation_parameters.multiphysics.cahn_hilliard_parameters,
+      (this->simulation_parameters.multiphysics.cahn_hilliard_parameters
```
Comment (AmishgaAlphonius) [inline source/solvers/cahn_hilliard.cc:50]: Nice ! :) 
but instead of repeating the if 3 times, wouldn't it be better to initialize a double once before and entering it as an argument.

File: source/solvers/cahn_hilliard.cc [hunk 5]
```diff
@@ -178,6 +184,8 @@ CahnHilliard<dim>::assemble_system_rhs()
   this->system_rhs = 0;
   setup_assemblers();
 
+
+
```
Comment (AmishgaAlphonius) [inline source/solvers/cahn_hilliard.cc:188]:
```suggestion
```

File: source/solvers/cahn_hilliard.cc [hunk 6]
```diff
@@ -278,6 +285,11 @@ CahnHilliard<dim>::attach_solution_to_output(DataOut<dim> &data_out)
   solution_names.push_back("phase_order");
   solution_names.push_back("chemical_potential");
 
+  std::vector<std::string> solution_names_filtered;
+  solution_names_filtered.push_back("phase_order_filtered");
+  solution_names_filtered.push_back("chemical_potential_filtered");
+
+
```
Comment (AmishgaAlphonius) [inline source/solvers/cahn_hilliard.cc:292]:
```suggestion
```

File: source/solvers/cahn_hilliard_assemblers.cc [hunk 7]
```diff
@@ -5,135 +5,86 @@
 #include <solvers/copy_data.h>
 
 
+
```
Comment (AmishgaAlphonius) [inline source/solvers/cahn_hilliard_assemblers.cc:8]:
```suggestion
```

File: source/solvers/cahn_hilliard_assemblers.cc (hunk 3 above)
Comment (AmishgaAlphonius) [inline source/solvers/cahn_hilliard_assemblers.cc:179]:
```suggestion
```

File: source/solvers/navier_stokes_scratch_data.cc [hunk 8]
```diff
@@ -482,38 +544,21 @@ NavierStokesScratchData<dim>::calculate_physical_properties()
             }
           else if (gather_cahn_hilliard && !gather_vof)
             {
-              // Blend the physical properties using the CahnHilliard field
+              //  Blend the physical properties using the CahnHilliard field
```
Comment (AmishgaAlphonius) [inline source/solvers/navier_stokes_scratch_data.cc:547]:
```suggestion
              // Blend the physical properties using the CahnHilliard field
```

File: source/solvers/cahn_hilliard_scratch_data.cc [hunk 9]
```diff
@@ -65,6 +65,66 @@ CahnHilliardScratchData<dim>::allocate()
...
+              const auto surface_tension_model =
+                properties_manager.get_surface_tension(material_interaction_id);
+              surface_tension_model->vector_value(fields, surface_tension);
+            }
+
+          // Gather surface tension
+          const auto material_interaction_id =
+            properties_manager.get_material_interaction_id(
+              material_interactions_type::fluid_fluid, 0, 1);
+
```
Comment (AmishgaAlphonius) [inline source/solvers/cahn_hilliard_scratch_data.cc:114]: For CH simulations, the number of material interactions is always 1 or over.

```suggestion
          const auto material_interaction_id =
            properties_manager.get_material_interaction_id(
              material_interactions_type::fluid_fluid, 0, 1);

          const auto surface_tension_model =
            properties_manager.get_surface_tension(material_interaction_id);
          surface_tension_model->vector_value(fields, surface_tension);
```

File: source/solvers/navier_stokes_base.cc [hunk 10]
```diff
@@ -1476,6 +1476,12 @@ NavierStokesBase<dim, VectorType, DofsType>::postprocess_fd(bool firstIter)
               // Calculate error on pressure for VOF simulations
               if (this->simulation_parameters.multiphysics.VOF)
                 this->error_table.add_value("error_pressure", error_pressure);
+
+              // Calculate error on pressure for Cahn-Hilliard simulations
+              if (this->simulation_parameters.multiphysics.cahn_hilliard)
+                {
+                  this->error_table.add_value("error_pressure", error_pressure);
+                }
```
Comment (AmishgaAlphonius) [inline source/solvers/navier_stokes_base.cc:1484]:
```suggestion
              // Calculate error on pressure for VOF simulations
              if (this->simulation_parameters.multiphysics.VOF ||
                  this->simulation_parameters.multiphysics.cahn_hilliard)
                this->error_table.add_value("error_pressure", error_pressure);
```

File: source/solvers/navier_stokes_base.cc [hunk 11]
```diff
@@ -1484,6 +1490,13 @@ NavierStokesBase<dim, VectorType, DofsType>::postprocess_fd(bool firstIter)
                           << std::setprecision(
                                simulation_control->get_log_precision())
                           << error_velocity << std::endl;
+              if (this->simulation_parameters.multiphysics.cahn_hilliard)
+                {
+                  this->pcout << "L2 error pressure : "
+                              << std::setprecision(
+                                   simulation_control->get_log_precision())
+                              << error_pressure << std::endl;
+                }
```
Comment (AmishgaAlphonius) [inline source/solvers/navier_stokes_base.cc:1499]: Why is the pressure only displayed for CH?

**Review by AmishgaAlphonius — CHANGES_REQUESTED** [review]
Great job @PierreLaurentinCS!!! It is nice to see everything coming together after so many months of hard work!!

I haven't finished my review. I'll look at the assemblers on Monday when I'll have the equations :)

File: applications_tests/lethe-fluid/mms2d_cahn_hilliard.prm [hunk 12]
```diff
@@ -40,7 +40,7 @@ subsection analytical solution
   set verbosity = verbose
 
   subsection cahn hilliard
-    set Function expression = sin(pi*x)*sin(pi*y); 1/0.0 #This is to make the test more robust on every configuration. Only the phase (first component) is being tested, the potential is not tested, the numerical solution converges to zero extremely fast.
+    set Function expression = -sin(pi*x)*sin(pi*y); 1/0.0 #This is to make the test more robust on every configuration. Only the phase (first component) is being tested, the potential is not tested, the numerical solution converges to zero extremely fast.
```
[the remark this answers is not in the document]
Comment (PierreLaurentinCS) [reply applications_tests/lethe-fluid/mms2d_cahn_hilliard.prm:43]: Since I am only interested in the phase order parameter, the chemical potential acts rather chaotic. To avoid the test to fail on other configurations, I set the analytical solution for the chemical potential to 1/0.0 so that it always print nan. By robust I meant that every configuration should be able to run this test, and I left the disclaimer because people might be surprised by the 1/0.0 term.

File: include/core/mobility_cahn_hilliard_model.h [hunk 13]
```diff
@@ -200,9 +179,13 @@ class MobilityCahnHilliardModelQuartic : public MobilityCahnHilliardModel
...
-    return mobility_cahn_hilliard_constant *
-           (1 - phase_order_cahn_hilliard * phase_order_cahn_hilliard) *
-           (1 - phase_order_cahn_hilliard * phase_order_cahn_hilliard);
+
+    if (std::abs(phase_order_cahn_hilliard) > 1)
+      return 0.0;
+    else
+      return mobility_cahn_hilliard_constant *
+             (1 - phase_order_cahn_hilliard * phase_order_cahn_hilliard) *
+             (1 - phase_order_cahn_hilliard * phase_order_cahn_hilliard);
```
[the remark this answers is not in the document]
Comment (PierreLaurentinCS) [reply include/core/mobility_cahn_hilliard_model.h:188]: agreed

File: include/core/mobility_cahn_hilliard_model.h [hunk 14]
```diff
@@ -236,8 +224,11 @@ class MobilityCahnHilliardModelQuartic : public MobilityCahnHilliardModel
   {
     const double &phase_order_cahn_hilliard =
       fields_value.at(field::phase_order_cahn_hilliard);
-    return -4 * phase_order_cahn_hilliard * mobility_cahn_hilliard_constant *
-           (1 - phase_order_cahn_hilliard * phase_order_cahn_hilliard);
+    if (std::abs(phase_order_cahn_hilliard) > 1)
+      return 0.0;
+    else
+      return -4 * phase_order_cahn_hilliard * mobility_cahn_hilliard_constant *
+             (1 - phase_order_cahn_hilliard * phase_order_cahn_hilliard);
```
[the remark this answers is not in the document]
Comment (PierreLaurentinCS) [reply include/core/mobility_cahn_hilliard_model.h:231]: agreed but it might make the return value less clear

File: source/solvers/cahn_hilliard.cc [hunk 15]
```diff
@@ -1163,14 +1206,16 @@ CahnHilliard<dim>::calculate_barycenter(const GlobalVectorType &solution,
...
-              volume += (1 - phase_cahn_hilliard_values[q]) * 0.5 * JxW;
-              barycenter_location += (1 - phase_cahn_hilliard_values[q]) * 0.5 *
-                                     quadrature_locations[q] * JxW;
-              barycenter_velocity += (1 - phase_cahn_hilliard_values[q]) * 0.5 *
-                                     velocity_values[q] * JxW;
+              volume += (1 - phase_values) * 0.25 * JxW;
+              barycenter_location +=
+                (1 - phase_values) * 0.25 * quadrature_locations[q] * JxW;
+              barycenter_velocity +=
+                (1 - phase_values) * 0.25 * velocity_values[q] * JxW;
```
[the remark this answers is not in the document]
Comment (PierreLaurentinCS) [reply source/solvers/cahn_hilliard.cc:1218]: You're right, it should be 0.5 and not 0.25, I did a change of variable, but it is not correct here. There is a factor because in the fluid 1, which have phase_value = -1, the integrand equals to 2. In the fluid 0 it is equal to 0

File: source/solvers/cahn_hilliard_assemblers.cc (hunk 2 above)
Comment (PierreLaurentinCS) [reply source/solvers/cahn_hilliard_assemblers.cc:19]: The smoothing parameter is always specified by the user, so it is kept in the parameters. Epsilon on the contrary can either be user specified or automatically determined. For the latter, the minimum cell diameter must be known : it is accessible through the minimum_cell_diameter function (GridTools class) used on the triangulation object, which, itself, is an attribute of the CahnHilliard class, but the the assembler.

Comment (PierreLaurentinCS) [reply source/solvers/cahn_hilliard_assemblers.cc:19]: Should I clarify this somewhere in the code ?

Comment (blaisb) [reply source/solvers/cahn_hilliard_assemblers.cc:19]: I would add a comment or something similar to make it specific why this is done this way

File: applications_tests/lethe-fluid/mms2d_cahn_hilliard.prm (hunk 12 above)
[the remark this answers is not in the document]
Comment (blaisb) [reply applications_tests/lethe-fluid/mms2d_cahn_hilliard.prm:43]: OKidoke

File: source/solvers/cahn_hilliard.cc (hunk 1 above)
Comment (PierreLaurentinCS) [reply source/solvers/cahn_hilliard.cc:1253]: agreed

File: applications_tests/lethe-fluid/gls_cahn_hilliard_clip_filter.prm [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+      set second fluid id = 1
+      # Surface tension
+      set surface tension model       = constant
+      set surface tension coefficient = 0
+      # Mobility for Cahn-Hilliard equations
+      set cahn hilliard mobility model    = constant
+      set cahn hilliard mobility constant = 0
+    end
+  end
+end
```
[the remark this answers is not in the document]
Comment (PierreLaurentinCS) [reply applications_tests/lethe-fluid/gls_cahn_hilliard_clip_filter.prm:89]: I agree on everything, except that the mobility default value is 1e-7 (which makes sense in most of problems contrary to 1!)

File: include/core/parameters_multiphysics.h [hunk 17]
```diff
@@ -63,6 +70,27 @@ namespace Parameters
...
+
+  /**
+   * @brief CahnHilliard_PhaseFilter - Defines the parameters for the phase filtration of CahnHilliard physics
+   */
+  struct CahnHilliard_PhaseFilter
+  {
+    // Type of filter
+    Parameters::FilterType type;
+
+    // $$\beta$$ value for the tanh filter
```
[the remark this answers is not in the document]
Comment (PierreLaurentinCS) [reply include/core/parameters_multiphysics.h:82]: Does $$ not work for Doxygen?

File: include/core/parameters_multiphysics.h [hunk 18]
```diff
@@ -26,9 +26,15 @@
...
 
+#include <core/dimensionality.h>
+#include <core/ib_particle.h>
+#include <core/multiphysics.h>
 #include <core/parameters.h>
+#include <core/utilities.h>
 
+#include <deal.II/base/conditional_ostream.h>
 #include <deal.II/base/parameter_handler.h>
+#include <deal.II/base/parsed_function.h>
```
[the remark this answers is not in the document]
Comment (PierreLaurentinCS) [reply include/core/parameters_multiphysics.h:37]: In my defence, it was here we the branch was forked from master (4500 years ago)

File: include/solvers/navier_stokes_scratch_data.h [hunk 19]
```diff
@@ -1166,15 +1188,21 @@ class NavierStokesScratchData
...
-  double                      epsilon;
-  double                      well_height;
-  double                      density_diff;
   bool                        gather_cahn_hilliard;
   unsigned int                n_dofs_cahn_hilliard;
   std::vector<double>         phase_order_cahn_hilliard_values;
+  std::vector<double>         filtered_phase_order_cahn_hilliard_values;
   std::vector<Tensor<1, dim>> phase_order_cahn_hilliard_gradients;
+  std::vector<Tensor<1, dim>>
+    filtered_phase_order_cahn_hilliard_gradients; // PAS INITIALISE ATTENTION
```
[the remark this answers is not in the document]
Comment (PierreLaurentinCS) [reply include/solvers/navier_stokes_scratch_data.h:1197]: everything related to this variable was removed because it is not used anymore

File: source/solvers/cahn_hilliard_assemblers.cc
[the remark this answers is not in the document]
Comment (PierreLaurentinCS) [reply source/solvers/cahn_hilliard_assemblers.cc:1]: I think no assembler have the copyright header

File: source/core/parameters_multiphysics.cc [hunk 20]
```diff
@@ -484,15 +484,89 @@ Parameters::VOF_PhaseFilter::parse_parameters(ParameterHandler &prm)
...
+  {
+    prm.declare_entry(
+      "type",
+      "none",
+      Patterns::Selection("none|clip|tanh"),
+      "CahnHilliard phase filtration type, "
+      "if <none> is selected, the phase won't be filtered"
+      "if <clip> is selected, the phase order values above 1 (respectively below -1) will be brought back to 1 (respectively -1)"
+      "if <tanh> is selected, the filtered phase will be a result of the "
+      "following function: \\alpha_f = \\tanh(\\beta\\alpha); "
```
[the remark this answers is not in the document]
Comment (PierreLaurentinCS) [reply source/core/parameters_multiphysics.cc:500]: I think so since, but the documentation is not so clear about it

File: source/core/parameters_multiphysics.cc [hunk 21]
```diff
@@ -518,18 +592,23 @@ Parameters::CahnHilliard::parse_parameters(ParameterHandler &prm)
...
+      prm.get_double("potential smoothing coefficient");
 
     prm.enter_subsection("epsilon");
     {
       const std::string op_epsilon = prm.get("method");
       if (op_epsilon == "automatic")
         {
-          epsilon_set_method = Parameters::EpsilonSetStrategy::automatic;
+          CahnHilliard::epsilon_set_method =
+            Parameters::EpsilonSetStrategy::automatic;
```
[the remark this answers is not in the document]
Comment (PierreLaurentinCS) [reply source/core/parameters_multiphysics.cc:606]: agreed, I will keep `method`

File: source/solvers/cahn_hilliard.cc (hunk 4 above)
Comment (PierreLaurentinCS) [reply source/solvers/cahn_hilliard.cc:50]: I thought so too but I want to keep the `setup_assemblers` function as "clean" as possible, and not have computations on the side. I don't think the three ifs have a big impact on performance, but I am open to discuss this

File: source/solvers/cahn_hilliard_scratch_data.cc (hunk 9 above)
Comment (PierreLaurentinCS) [reply source/solvers/cahn_hilliard_scratch_data.cc:114]: agreed

File: source/solvers/navier_stokes_base.cc (hunk 11 above)
Comment (PierreLaurentinCS) [reply source/solvers/navier_stokes_base.cc:1499]: I thought, since it was not implemented before that there was a specific reason for it, so I left it only for the Cahn-Hilliard simulations

**Review by blaisb — CHANGES_REQUESTED** [review]
Moar comments

File: source/core/parameters_multiphysics.cc (hunk 20 above)
[the remark this answers is not in the document]
Comment (blaisb) [reply source/core/parameters_multiphysics.cc:500]: In all cases this will never be printed in LaTeX format This is just the default comment used in the template parameter generator.

File: source/core/parameters_multiphysics.cc [hunk 22]
```diff
@@ -484,15 +484,88 @@ Parameters::VOF_PhaseFilter::parse_parameters(ParameterHandler &prm)
...
+    prm.declare_entry(
+      "type",
+      "none",
+      Patterns::Selection("none|clip|tanh"),
+      "CahnHilliard phase filtration type, "
+      "if <none> is selected, the phase won't be filtered"
+      "if <clip> is selected, the phase order values above 1 (respectively below -1) will be brought back to 1 (respectively -1)"
+      "if <tanh> is selected, the filtered phase will be a result of the "
+      "following function: \\alpha_f = \\tanh(\\beta\\alpha); "
+      "where \\beta is a parameter influencing the interface thickness that "
```
Comment (blaisb) [inline source/core/parameters_multiphysics.cc:501]:
```suggestion
      "where beta is a parameter influencing the interface thickness that "
```

File: applications_tests/lethe-fluid/gls_cahn_hilliard_clip_filter.prm (hunk 16 above)
[the remark this answers is not in the document]
Comment (AmishgaAlphonius) [reply applications_tests/lethe-fluid/gls_cahn_hilliard_clip_filter.prm]: You're right, but this must be changed in documentation :)

File: examples/multiphysics/rising-bubble-chns/rising-bubble-chns.prm [hunk 23]
```diff
@@ -59,164 +95,133 @@ end
...
 
 subsection linear solver
   subsection fluid dynamics
     set verbosity                             = verbose
     set method                                = gmres
-    set max iters                             = 20000
-    set relative residual                     = 1e-3
-    set minimum residual                      = 1e-10
-    set preconditioner                        = ilu
+    set max iters                             = 8000
```
[the remark this answers is not in the document]
Comment (PierreLaurentinCS) [reply examples/multiphysics/rising-bubble-chns/rising-bubble-chns.prm:207]: no

File: source/solvers/navier_stokes_base.cc (hunk 11 above)
Comment (blaisb) [reply source/solvers/navier_stokes_base.cc:1495]: The issue is that for the VOF solver the pressure is sometimes erratic (it will translate by a constant) so generally pressure in VOF is not a very portable variable. I'M not sure if it's a good idea to enable it in a test. We can try and if it is portable we will keep it, if no, we can remove it in the future.

**Review by blaisb — CHANGES_REQUESTED** [review]
A little bit moar.

File: source/solvers/cahn_hilliard.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (PierreLaurentinCS) [reply source/solvers/cahn_hilliard.cc]: Strangely, it did not have any noticeable effect on the examples which were using this functionality after double checking

**Review by AmishgaAlphonius — APPROVED** [review]
Only some minor comments :)
Great work again!

**Review by blaisb — APPROVED** [review]
